pycqed/measurement/waveform_control/pulse_library.py

```diff
@@ -113,6 +113,188 @@ def hashables(self, tstart, channel):
+        assert len(amplitudes) == len(channels)
+        for l, a in zip(lengths, amplitudes):
+            assert len(l) == len(a)
+
+    @property
+    def length(self):
+        max_len = 0
+        for channel_lengths in self.lengths:
+            max_len = max(max_len, np.sum(channel_lengths))
+        return max_len
+
+    def _check_dimensions(self):
+        assert len(self.lengths) == len(self.channels)
+        assert len(self.amplitudes) == len(self.channels)
+        for chan_lens, chan_amps in zip(self.lengths, self.amplitudes):
+            assert len(chan_lens) == len(chan_amps)
+
+    def chan_wf(self, channel, t):
+        t0 = self.algorithm_time()
+        idx = self.channels.index(channel)
+        wave = np.zeros_like(t)
+
+        if self.gaussian_filter_sigma > 0:
+            timescale = 1 / (np.sqrt(2) * self.gaussian_filter_sigma)
+        else:
+            timescale = 0
+
+        for seg_len, seg_amp in zip(self.lengths[idx], self.amplitudes[idx]):
+            t1 = t0 + seg_len
+            if self.gaussian_filter_sigma > 0:
+                wave += 0.5 * seg_amp * (sp.special.erf((t - t0) * timescale) -
+                                         sp.special.erf((t - t1) * timescale))
+            else:
+                wave += seg_amp * (t >= t0) * (t < t1)
+            t0 = t1
+        return wave
+
+    def hashables(self, tstart, channel):
+        if channel not in self.channels:
+            return []
+        hashlist = [type(self), self.algorithm_time() - tstart]
+        idx = self.channels.index(channel)
+        chan_lens = self.lengths[idx]
+        chan_amps = self.amplitudes[idx]
+        hashlist += [len(chan_lens)]
+        hashlist += list(chan_lens.copy())
+        hashlist += list(chan_amps.copy())
+        hashlist += [self.gaussian_filter_sigma]
+        return hashlist
+
+
+class NZTransitionControlledPulse(GaussianFilteredPiecewiseConstPulse):
+    """
+    A zero-area pulse shape that allows to control the accumulated phase when
+    transitioning from the first pulse half to the second pulse half, by having
+    an additional, low-amplitude segment between the two main pulse-halves.
+
+    The zero area is achieved by adjusting the lengths for the intermediate
+    pulses.
+    """
+
+    def __init__(self, name, element_name, **kw):
+        GaussianFilteredPiecewiseConstPulse.__init__(
+            self, name, element_name, [], [], [], **kw)
+        pulse.Pulse.__init__(self, name, element_name, **kw)
+
+        for par in ['amplitude', 'amplitude2', 'amplitude_offset',
+                    'amplitude_offset2', 'pulse_length', 'trans_amplitude',
+                    'trans_amplitude2', 'trans_length', 'buffer_length_start',
+                    'buffer_length_end', 'channel_relative_delay']:
+            self.add_value_change_callback(par, self._update_lengths_amps)
+        for par in ['channel', 'channel2']:
+            self.add_value_change_callback(par, self._update_channels)
```

Comment:
In pull request #37, we removed the __call__ function from all pulses. As far as I understood back then, the argument was: pulse parameters are chosen before the actual pulse object is created by the init of UnresolvedPulse(), and there is no need to modify them afterwards. Now, this callback procedure is introduced to handle parameter changes after the creation of this pulse (now by manually assigning a parameter instead of the original concept of calling the pulse to change parameters). If we think that there is a use-case for changing pulse parameters after object creation, wouldn't it make sense to introduce a more general framework that allows late parameter changes (an appropriate handling of them) for all kinds of pulses?

Reply:
I think the parameters are changed correctly for other pulses as the waveforms are generated from the parameters that the user modifies.

In this case the pulse shape is internally stored in a more general way (GaussianFilteredPiecewiseConstPulse) than the parameters that the user has access to. The callbacks are needed to convert from the interface-parametrization to the internal data storage parametrization.
I think this is still a rather special pulse class in that sense.

I wouldn't like to enforce the __call__() syntax for all pulses, since it is much less intuitive than just setting corresponding the pulse parameters and lead to a lot of duplicate code.

Reply:
OK, makes sense. But maybe the helper function add_value_change_callback could be defined in pulse.py, then every pulse subclass that needs callbacks could just call add_value_change_callback in the subclass init. Another use case for this will be the C-ARB gate, where we indeed had the issue that the desired cphase could not be updated anymore after creating the pulse object.

Or alternatively, the pulse base class could have a generic setter function monitoring all parameter changes, and a property called e.g. self.callback_dict. Then, the child classes would just add entries of the form {'param_name': function_handle} to this dict. The generic setter function in the pulse base class would then check whether there is a callback in this dict for the parameter that is set.

Reply:
Looking at the code again, the callback thing did not make so much sense actually.

I will now call `_update_lengths_amps` and `_update_channels` only when generating waveforms.

Reply:
So I will delete the add_value_change_callback